man/man1/logtalk_user_setup.1
.TH logtalk_user_setup 1 "June 10, 2010" "Logtalk 2.44.1" "Logtalk Documentation"

.SH NAME
logtalk_user_setup \-- Logtalk user setup script

.SH SYNOPSIS
.B logtalk_user_setup

.SH DESCRIPTION
\f3logtalk_user_setup\f1 is a Logtalk user setup script for copying the Logtalk user-modifiable files and directories to the location pointed by the environment variable LOGTALKUSER (defaults to "~/logtalk"). When the Logtalk user directory already exists, a backup is made to a directory whose name starts with "logtalk-backup-" and a new Logtalk user directory is created.

.SH ENVIRONMENT
.TP
.B LOGTALKHOME
Specifies the location of the Logtalk installation.
.TP
.B LOGTALKUSER
Specifies the location of the Logtalk user directory.

.SH "SEE ALSO"
The Logtalk website at http://logtalk.org/
.PP
The Logtalk User and Reference Manuals at $LOGTALKHOME/manuals/index.html
.PP
logtalk_version_select(1),\ logtalk_backend_select(1)

.SH AUTHOR
Paulo Moura <pmoura@logtalk.org>

.SH COPYRIGHT
Copyright (c) 1998-2012 Paulo Moura.


RELEASE_NOTES
=================================================================
Logtalk - Object oriented extension to Prolog
Release 2.28.0

Copyright (c) 1998-2006 Paulo Moura.  All Rights Reserved.
=================================================================


RELEASE NOTES
=============


2.28.0 - June ??, 2006

	Added support for multi-threading programming on selected Prolog compilers.
	Added a new object directive, threaded/0, and four new built-in predicates,
	threaded_call/1-2 and threaded_exit/1-2. Added a new set of exemples, 
	"threads".

	Updated the config file for the forthcoming SICStus Prolog 4.0 version 
	to work in the current beta version. Modified the "make_sicstus.lgt.js" 
	script to work with both SICStus Prolog 3.12 and 4.0 versions.

	Updated the config file for YAP to workaround a problem on the Windows 
	version where testing for directory existence fails if the path ends with 
	a slash (or a double backslash); this problem prevents the use of library 
	notation to load source files.

	Dropped support for versions of XSB older than 3.0.

	Extended "parametric" and "metainterpreters" examples.

	Update all the JScript scripts to check if a compatible version of WSH is 
	installed and to workaround some problems with spaces in file paths.

	Added Inno Setup GUI Windows installer script.


2.27.1 - March 27, 2006

	Allow calls to the built-in method parameter/2 with the first argument 
	not instantiated at compile time. Detect and report as a compilation 
	error parameter/2 indexes that are out of range.

	Optimized generation of predicate definition and declaration linking 
	clauses, resulting in small lookup performance improvements and in space 
	savings for the Prolog code generated when compiling Logtalk entities.

	Many minor code and documentation improvements to the Logtalk compiler.

	Added an object version of the tokenizer described in the Michael 
	Covington's paper "Tokenization using DCG Rules" to the "dcgs" example.

	Improved integration code for Qu-Prolog and Logtalk ("configs/qphook.ql").

	Improved library hierarchy methods that return lists of objects in order 
	to avoid duplicated elements (library objects "proto_hierarchy.lgt" and 
	"class_hierarchy.lgt"). Added new methods extension/1 and extensions/1 to 
	the library object "proto_hierarchy.lgt".

	Documented the concept of "parametric object proxy". Added a new example, 
	"proxies", illustrating the use of parametric object proxies.

	Added support for code completion and for listing entity names on the 
	symbol pop-up menu to the MacOS X TextMate text editor.

	Updated the "cplgtdirs.*" scripts to also create an alias/shortcut to the 
	"wenv" directory.

	Renamed the alternative compilation and documentation directory names on 
	all config files to be compatible across operating-systems. Removed unused
	predicate '$lgt_reverse'/2 from all config files.

	Updated HTML manuals "print.css" CSS file for compatibility with the 
	latest version of CSSToXSLFO.


2.27.0 - February 9, 2006

	Improved performance for local calls to the built-in methods phrase/2-3.
	Allow the built-in methods phrase/2-3 to accept both partial and proper 
	lists of terminals.

	Improved grammar rule translator. Report calls to undefined non-terminals 
	when translating grammar rules.

	Added support for declaring grammar rule non-terminal aliases using the 
	alias/3 directive.

	Added a new predicate property, non_terminal/1, for predicates resulting 
	from the compilation of grammar rule non-terminals.

	Improved support for the built-in method expand_term/2 in order to allow 
	bypassing of the default Logtalk grammar rule translator by defining 
	clauses for the term_expansion/2 predicate. Added a "term_expansionp" 
	protocol to the Logtalk library. Added a new example, "expansion", 
	illustrating the use of the term_expansion/2 predicate.

	Added a new compiler flag, hook/1, allowing the specification of a 
	compiler hook that is called for which term read from a source file.
	Added a simple example, "hooks", of using the Logtalk compiler hook 
	to expand author abbreviations in info/1 directives into full names 
	and email addresses.

	Added support for XSB 3.x to the runtime error handler. Updated the XSB 
	3.x config file with declarations for multi-threading meta-predicates.
	
	Removed a few choice-points when checking validity of entity directives.

	Added two new objects, list(Type) and set(Type), to the standard library 
	supporting the validation of lists and ordered sets whose elements are 
	restricted to a single type.

	Added a new DCG example for solving enigmas encoded using a cellphone 
	keypad.

	Added a missing library dependency to the "puzzles" example SCRIPT file.

	Removed the experimental "systemp.lgt" protocol from the list of files 
	loaded by the "library/all_loader.lgt" loader utility files to avoid 
	compilation errors on some Prolog systems.

	Corrected a bug that prevented dynamic creation of categories using the 
	built-in predicate create_category/4.

	Corrected a bug in the reporting of singleton variables, which failed to 
	write an accurate message for facts and grammar rules.

	Corrected a bug in passing the correct calling context ("this") when 
	processing meta-calls in objects.

	Corrected a bug in scope checking with local calls to reflection and 
	database methods.

	Corrected a bug in checking the validity of the arguments of the op/3 
	directive appearing inside entities.

	Added predicates for testing if a term is a partial list or a proper list 
	to all config files.

	Added a definition for missing open/4 ISO Prolog predicate to the config 
	file of Bin Prolog.

	Added a workaround for a B-Prolog bug to this compiler config file.


2.26.2 - December 20, 2005

	Improved error-checking for the Logtalk compiler and for the grammar rule 
	translator.

	Small performance improvements for message sending and for the built-in 
	database methods.


scripts/logtalk_user_setup.js
// ================================================================
// Logtalk - Open source object-oriented logic programming language
// Release 2.44.1
//
// Copyright (c) 1998-2012 Paulo Moura. All Rights Reserved.
// 
// Logtalk is free software. You can redistribute it and/or modify
// it under the terms of the Artistic License 2.0 as published by 
// the The Perl Foundation.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// Artistic License 2.0 for more details. A copy of the license is 
// provided in the "LICENSE.txt" file.
// ================================================================

if (ScriptEngineMajorVersion() < 5 || ScriptEngineMajorVersion() == 5 && ScriptEngineMinorVersion() < 6) {
	WScript.Echo('Error! WSH 5.6 or later version needed for running this script.');
	WScript.Quit(1);
}

var WshShell = new ActiveXObject("WScript.Shell");

var WshSystemEnv = WshShell.Environment("SYSTEM");
var WshUserEnv = WshShell.Environment("USER");
var logtalk_home;
var logtalk_user;

if (WshSystemEnv.Item("LOGTALKHOME"))
	logtalk_home = WshSystemEnv.Item("LOGTALKHOME");
else if (WshUserEnv.Item("LOGTALKHOME"))
	logtalk_home = WshUserEnv.Item("LOGTALKHOME")
else {
	WScript.Echo("The environment variable LOGTALKHOME must be defined first!");
	usage_help();
	WScript.Quit(1);
}

var FSObject = new ActiveXObject("Scripting.FileSystemObject");

if (!FSObject.FolderExists(logtalk_home)) {
	WScript.Echo("The environment variable LOGTALKHOME points to a non-existing directory!");
	WScript.Echo("Its current value is: %LOGTALKHOME%");
	WScript.Echo("The variable must be set to your Logtalk installation directory!");
	WScript.Echo("");
	usage_help();
	WScript.Quit(1);
}

if (WScript.Arguments.Unnamed.Length > 0)
	usage_help();

if (WshUserEnv.Item("LOGTALKUSER"))
	logtalk_user = WshUserEnv.Item("LOGTALKUSER");
else {
	logtalk_user = WshShell.SpecialFolders("MyDocuments") + "\\logtalk";
	WshUserEnv.Item("LOGTALKUSER") = WshShell.SpecialFolders("MyDocuments") + "\\logtalk";
	WScript.Echo("Defined user environment variable LOGTALKUSER.");
	WScript.Echo("");
}

if (FSObject.FolderExists(logtalk_user)) {
	var today = new Date();
	var year  = today.getFullYear();
	var month = today.getMonth() + 1;
	if (month < 10)
		month = "0" + month;
	var day = today.getDate();
	if (day < 10)
        day = "0" + day;
	var hours = today.getHours();
	if (hours < 10)
        hours = "0" + hours;
	var mins = today.getMinutes();
	if (mins < 10)
        mins = "0" + mins;
	var secs = today.getSeconds();
	if (secs < 10)
		secs = "0" + secs;
	date = year + "-" + month + "-" + day + "-" + hours + mins + secs;
	FSObject.MoveFolder(logtalk_user, logtalk_user + "-backup-" + date);
	WScript.Echo("Created a backup of the existing \"\%LOGTALKUSER\%\" directory:");
	WScript.Echo("");
	WScript.Echo("  " + logtalk_user + "-backup-" + date);
	WScript.Echo("");
	WScript.Echo("Creating a new \"\%LOGTALKUSER\%\" directory:");
	WScript.Echo("");
	WScript.Echo("  " + logtalk_user);
	WScript.Echo("");
	FSObject.CreateFolder(logtalk_user);
	if (FSObject.FileExists(logtalk_user + "-backup-" + date + "\\settings.lgt")) {
		FSObject.CopyFile(logtalk_user + "-backup-" + date + "\\settings.lgt", logtalk_user + "\\settings.lgt");
		WScript.Echo("Copied your old \"settings.lgt\" file to the new \"\%LOGTALKUSER\%\" directory.");
		WScript.Echo("The file \"settings-pristine.lgt\" file contains a pristine copy of the ");
		WScript.Echo("\"settings.lgt\" file distributed with the currently installed Logtalk");
		WScript.Echo("version. Review this file for possible settings files update instructions.");
	}
	WScript.Echo("");
} else {
	WScript.Echo("Creating a new \"\%LOGTALKUSER\%\" directory:");
	WScript.Echo("");
	WScript.Echo("  " + logtalk_user);
	WScript.Echo("");
	FSObject.CreateFolder(logtalk_user);
}

WScript.Echo("Copying Logtalk files and directories...");
FSObject.CopyFolder(logtalk_home + "\\contributions", logtalk_user + "\\contributions");
FSObject.CopyFolder(logtalk_home + "\\examples", logtalk_user + "\\examples");
FSObject.CopyFolder(logtalk_home + "\\libpaths", logtalk_user + "\\libpaths");
FSObject.CopyFolder(logtalk_home + "\\library", logtalk_user + "\\library");
FSObject.CopyFolder(logtalk_home + "\\xml", logtalk_user + "\\xml");
if (FSObject.FileExists(logtalk_user + "\\settings.lgt"))
	FSObject.CopyFile(logtalk_home + "\\settings.lgt", logtalk_user + "\\settings-pristine.lgt");
else
	FSObject.CopyFile(logtalk_home + "\\settings.lgt", logtalk_user + "\\settings.lgt");
FSObject.CopyFile(logtalk_home + "\\VERSION.txt", logtalk_user + "\\VERSION.txt");

FSObject.DeleteFile(logtalk_user + "\\xml\\lgt2*.*");
FSObject.DeleteFile(logtalk_user + "\\xml\\logtalk.dtd");
FSObject.DeleteFile(logtalk_user + "\\xml\\logtalk.xsd");

var link = WshShell.CreateShortcut(logtalk_user + "\\configs.lnk");
link.Description = "Shortcut to Logtalk config files";
link.TargetPath = logtalk_home + "\\configs";
link.Save();

link = WshShell.CreateShortcut(logtalk_user + "\\manuals.lnk");
link.Description = "Shortcut to Logtalk documentation";
link.TargetPath = logtalk_home + "\\manuals";
link.Save();

link = WshShell.CreateShortcut(logtalk_user + "\\wenv.lnk");
link.Description = "Shortcut to Logtalk text editing support";
link.TargetPath = logtalk_home + "\\wenv";
link.Save();

link = WshShell.CreateShortcut(logtalk_user + "\\xml\\lgt2html.js.lnk");
link.Description = "Shortcut to lgt2html script";
link.TargetPath = logtalk_home + "\\xml\\lgt2html.js";
link.Save();

link = WshShell.CreateShortcut(logtalk_user + "\\xml\\lgt2pdf.js.lnk");
link.Description = "Shortcut to lgt2pdf script";
link.TargetPath = logtalk_home + "\\xml\\lgt2pdf.js";
link.Save();

link = WshShell.CreateShortcut(logtalk_user + "\\xml\\lgt2xml.js.lnk");
link.Description = "Shortcut to lgt2xml script";
link.TargetPath = logtalk_home + "\\xml\\lgt2xml.js";
link.Save();

link = WshShell.CreateShortcut(logtalk_user + "\\xml\\lgt2txt.js.lnk");
link.Description = "Shortcut to lgt2txt script";
link.TargetPath = logtalk_home + "\\xml\\lgt2txt.js";
link.Save();

link = WshShell.CreateShortcut(logtalk_user + "\\xml\\logtalk.dtd.lnk");
link.Description = "Shortcut to Logtalk DTD";
link.TargetPath = logtalk_home + "\\xml\\logtalk.dtd";
link.Save();

link = WshShell.CreateShortcut(logtalk_user + "\\xml\\logtalk.rng.lnk");
link.Description = "Shortcut to Logtalk RELAX NG Schema";
link.TargetPath = logtalk_home + "\\xml\\logtalk.rng";
link.Save();

link = WshShell.CreateShortcut(logtalk_user + "\\xml\\logtalk.xsd.lnk");
link.Description = "Shortcut to Logtalk XML Schema";
link.TargetPath = logtalk_home + "\\xml\\logtalk.xsd";
link.Save();

WScript.Echo("Finished copying Logtalk files and directories.");
WScript.Echo("");
WScript.Echo("You may want to customize the default Logtalk compiler flags by editing");
WScript.Echo("the \"settings.lgt\" file found in the directory \"\%LOGTALKUSER\%\". For more");
WScript.Echo("information on customizing Logtalk and your working environment, consult");
WScript.Echo("the \"\%LOGTALKUSER\%\\CUSTOMIZE.txt\" file for details.");
WScript.Echo("");

WScript.Quit(0);

function usage_help() {
	WScript.Echo("");
	WScript.Echo("This script copies the Logtalk user-modifiable files and directories");
	WScript.Echo("to the user home directory. The location can be set by the environment");
	WScript.Echo("variable \"\%LOGTALKUSER\%\" (defaults to \"MyDocuments\\logtalk\" when the");
	WScript.Echo("variable is not defined)");
	WScript.Echo("");
	WScript.Echo("Usage:");
	WScript.Echo("  " + WScript.ScriptName + " help");
	WScript.Echo("  " + WScript.ScriptName);
	WScript.Echo("");
	WScript.Quit(1);
}
